core/src/block_creation_loop.rs

```diff
@@ -238,40 +247,75 @@ fn start_loop(config: BlockCreationLoopConfig) {
             );
         }
 
-        // Wait for the voting loop to notify us, draining all pending messages and keeping the highest slot
+        // Race between parent ready notification and optimistic parent events.
+        // Parent ready is checked first and has priority if both channels are ready.
+        let window_source = select_biased! {
+            recv(ctx.leader_window_info_receiver) -> msg => {
+                msg.ok().map(ParentSource::ParentReady)
+            },
+            recv(optimistic_parent_receiver) -> msg => {
+                msg.ok().map(ParentSource::OptimisticParent)
+            },
+            default(Duration::from_secs(1)) => None,
```

Comment:
think we can just `continue` here
```suggestion
            default(Duration::from_secs(1)) => continue,
```